# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [1]:
# import numpy and pandas
import pandas as pd
import numpy as np
from scipy.stats import trim_mean, mode, skew, gaussian_kde, pearsonr, spearmanr, beta
from statsmodels.stats.weightstats import ztest as ztest

from scipy.stats import ttest_ind, norm, t
from scipy.stats import f_oneway
from scipy.stats import sem

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the City of Chicago. We will start by loading the dataset and examining its contents

In [3]:
# Run this code:
salaries = pd.read_csv('H:\My Drive\IRONHACK\WEEK_4\DAY_3\lab-hypothesis-testing-en\data\Current_Employee_Names__Salaries__and_Position_Titles.csv')

In [4]:
salaries.shape


(33183, 8)

In [5]:
salaries.columns


Index(['Name', 'Job Titles', 'Department', 'Full or Part-Time',
       'Salary or Hourly', 'Typical Hours', 'Annual Salary', 'Hourly Rate'],
      dtype='object')

Examine the `salaries` dataset using the `head` function below.

In [6]:
# Your code here
salaries.head()


Name                              Job Titles  \
0    AARON,  JEFFERY M                                SERGEANT   
1      AARON,  KARINA   POLICE OFFICER (ASSIGNED AS DETECTIVE)   
2  AARON,  KIMBERLEI R                CHIEF CONTRACT EXPEDITER   
3  ABAD JR,  VICENTE M                       CIVIL ENGINEER IV   
4    ABASCAL,  REECE E             TRAFFIC CONTROL AIDE-HOURLY   

         Department Full or Part-Time Salary or Hourly  Typical Hours  \
0            POLICE                 F           Salary            NaN   
1            POLICE                 F           Salary            NaN   
2  GENERAL SERVICES                 F           Salary            NaN   
3       WATER MGMNT                 F           Salary            NaN   
4              OEMC                 P           Hourly           20.0   

   Annual Salary  Hourly Rate  
0       101442.0          NaN  
1        94122.0          NaN  
2       101592.0          NaN  
3       110064.0          NaN  
4            NaN        19.86

In [7]:
salaries.tail()


Name               Job Titles Department  \
33178  ZYLINSKA,  KATARZYNA            POLICE OFFICER     POLICE   
33179     ZYMANTAS,  LAURA C           POLICE OFFICER     POLICE   
33180      ZYMANTAS,  MARK E           POLICE OFFICER     POLICE   
33181    ZYRKOWSKI,  CARLO E           POLICE OFFICER     POLICE   
33182   ZYSKOWSKI,  DARIUSZ   CHIEF DATA BASE ANALYST       DoIT   

      Full or Part-Time Salary or Hourly  Typical Hours  Annual Salary  \
33178                 F           Salary            NaN        72510.0   
33179                 F           Salary            NaN        48078.0   
33180                 F           Salary            NaN        90024.0   
33181                 F           Salary            NaN        93354.0   
33182                 F           Salary            NaN       115932.0   

       Hourly Rate  
33178          NaN  
33179          NaN  
33180          NaN  
33181          NaN  
33182          NaN

In [8]:
salaries.dtypes


Name                  object
Job Titles            object
Department            object
Full or Part-Time     object
Salary or Hourly      object
Typical Hours        float64
Annual Salary        float64
Hourly Rate          float64
dtype: object

In [9]:
salaries.isna().sum()


Name                     0
Job Titles               0
Department               0
Full or Part-Time        0
Salary or Hourly         0
Typical Hours        25161
Annual Salary         8022
Hourly Rate          25161
dtype: int64

In [ ]:
# numerical summary
salaries.describe()


Typical Hours  Annual Salary  Hourly Rate
count    8022.000000   25161.000000  8022.000000
mean       34.507604   86786.999790    32.788558
std         9.252077   21041.354602    12.112573
min        10.000000    7200.000000     2.650000
25%        20.000000   76266.000000    21.200000
50%        40.000000   90024.000000    35.600000
75%        40.000000   96060.000000    40.200000
max        40.000000  300000.000000   109.000000

In [11]:
# categorical
salaries.describe(include='object')


Name      Job Titles Department Full or Part-Time  \
count                33183           33183      33183             33183   
unique               32880            1111         35                 2   
top     HERNANDEZ,  JUAN C  POLICE OFFICER     POLICE                 F   
freq                     4            9520      13414             31090   

       Salary or Hourly  
count             33183  
unique                2  
top              Salary  
freq              25161

In [13]:
#checking unique values
salaries['Department'].nunique()
salaries['Department'].unique()

array(['POLICE', 'GENERAL SERVICES', 'WATER MGMNT', 'OEMC',
       'CITY COUNCIL', 'AVIATION', 'STREETS & SAN', 'FIRE',
       'FAMILY & SUPPORT', 'PUBLIC LIBRARY', 'TRANSPORTN',
       "MAYOR'S OFFICE", 'HEALTH', 'BUSINESS AFFAIRS', 'LAW', 'FINANCE',
       'CULTURAL AFFAIRS', 'COMMUNITY DEVELOPMENT', 'PROCUREMENT',
       'BUILDINGS', 'ANIMAL CONTRL', 'CITY CLERK', 'BOARD OF ELECTION',
       'DISABILITIES', 'HUMAN RESOURCES', 'DoIT', 'BUDGET & MGMT',
       'TREASURER', 'INSPECTOR GEN', 'HUMAN RELATIONS', 'COPA',
       'BOARD OF ETHICS', 'POLICE BOARD', 'ADMIN HEARNG',
       'LICENSE APPL COMM'], dtype=object)

In [14]:
salaries['Full or Part-Time'].value_counts()

Full or Part-Time
F    31090
P     2093
Name: count, dtype: int64

In [15]:
salaries['Job Titles'].value_counts().head(10)

Job Titles
POLICE OFFICER                            9520
FIREFIGHTER-EMT                           1473
SERGEANT                                  1202
POOL MOTOR TRUCK DRIVER                   1055
POLICE OFFICER (ASSIGNED AS DETECTIVE)     989
MOTOR TRUCK DRIVER                         705
SANITATION LABORER                         625
FIREFIGHTER                                487
CONSTRUCTION LABORER                       465
CROSSING GUARD                             451
Name: count, dtype: int64

In [17]:
salaries['Annual Salary'].dtype

dtype('float64')

In [18]:
# just make sure it's float
salaries['Annual Salary'] = salaries['Annual Salary'].astype(float)


In [19]:
salaries['Annual Salary'].head()


0    101442.0
1     94122.0
2    101592.0
3    110064.0
4         NaN
Name: Annual Salary, dtype: float64

The Annual Salary column is already stored as a numeric (float64) data type.
Therefore, string operations using .str are not applicable and not required.

In [20]:
salaries['Annual Salary'].describe()


count     25161.000000
mean      86786.999790
std       21041.354602
min        7200.000000
25%       76266.000000
50%       90024.000000
75%       96060.000000
max      300000.000000
Name: Annual Salary, dtype: float64

In [21]:
salaries['Annual Salary'].isna().sum()


np.int64(8022)

# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of all hourly workers is significantly different from $30/hr. Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [22]:
# Your code here
# STEP 1 - Import the correct test from SciPy

from scipy.stats import ttest_1samp


In [24]:
# step 2 - Filter hourly workers

hourly_workers = salaries[salaries['Salary or Hourly'] == 'Hourly']

In [25]:
# STEP 3 - Extract hourly wages (drop missing values!)

hourly_wages = hourly_workers['Hourly Rate'].dropna()


In [26]:
# STEP 4 - Perform the one-sample t-test

t_stat, p_value = ttest_1samp(hourly_wages, popmean=30)

In [27]:
# STEP 5 - View results

t_stat, p_value

(np.float64(20.6198057854942), np.float64(4.323024048623005e-92))

In [28]:
# STEP 6 - Decision rule (α = 0.05)
alpha = 0.05

if p_value < alpha:
    print("Reject the null hypothesis")
else:
    print("Fail to reject the null hypothesis")


Reject the null hypothesis


A two-sided one-sample t-test was conducted to determine whether the mean hourly wage of hourly workers differs from $30 per hour.
The test yielded a t-statistic of 20.62 and a p-value of 4.32 × 10⁻⁹².
Since the p-value is far below the significance level of 0.05, we reject the null hypothesis.
This indicates that the mean hourly wage of hourly workers is significantly different from $30 per hour.

# Challenge 3 - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

The confidence interval is computed in SciPy using the `t.interval` function. You can read more about this function [here](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.t.html).

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level, number of rows - 1 for degrees of freedom, the mean of the sample for the location parameter and the standard error for the scale. The standard error can be computed using [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html) function in SciPy.

In [29]:
# Your code here
# STEP 1 -  Import the required SciPy functions

from scipy.stats import t, sem

#t.interval → builds the confidence interval
#sem → computes the standard error of the mean

In [30]:
# STEP 2 - Reuse the hourly wage data
# From Challenge 2 — no need to redo filtering

hourly_workers = salaries[salaries['Salary or Hourly'] == 'Hourly']
hourly_wages = hourly_workers['Hourly Rate'].dropna()


In [31]:
# STEP 3 - Compute summary statistics

mean_wage = hourly_wages.mean()
std_error = sem(hourly_wages)
n = len(hourly_wages)
df = n - 1


In [32]:
# STEP 4 - Construct the 95% confidence interval

confidence_interval = t.interval(
    confidence=0.95,
    df=df,
    loc=mean_wage,
    scale=std_error
)

confidence_interval


(np.float64(32.52345834488425), np.float64(33.05365708767623))

A 95% confidence interval for the mean hourly wage of hourly workers is [$32.52,\ $33.05].
This means we are 95% confident that the true population mean hourly wage lies between $32.52 and $33.05.

# Challenge 4 - Hypothesis Tests of Proportions

Another type of one sample test is a hypothesis test of proportions. In this test, we examine whether the proportion of a group in our sample is significantly different than a fraction. 

You can read more about one sample proportion tests [here](http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/SAS/SAS6-CategoricalData/SAS6-CategoricalData2.html).

In the cell below, use the `proportions_ztest` function from `statsmodels` to perform a hypothesis test that will determine whether the number of hourly workers in the City of Chicago is significantly different from 25% at the 95% confidence level.

In [33]:
# Your code here

# Goal of the test: Determine whether the proportion of hourly workers in the City of Chicago is significantly different from 25%.

# Hypotheses (write this in your lab!)
# - Null hypothesis (H₀): p = 0.25
# - Alternative hypothesis (H₁): p ≠ 0.25
  # - (two-sided test, α = 0.05)

# STEP 1 - Import the correct function

from statsmodels.stats.proportion import proportions_ztest



Why z-test?

Large sample size
Proportion-based inference
Standard approach for one-sample proportion tests

In [34]:
# STEP 2 - Define successes and sample size

# Here:
 # Success = employee is an hourly worker
 # Sample size = total number of employees

total_employees = len(salaries)
hourly_count = (salaries['Salary or Hourly'] == 'Hourly').sum()


In [35]:
# STEP 3 - Perform the one-sample proportion z-test

z_stat, p_value = proportions_ztest(
    count=hourly_count,
    nobs=total_employees,
    value=0.25,
    alternative='two-sided'
)


In [36]:
# STEP 4 - View results

z_stat, p_value

(np.float64(-3.5099964213703005), np.float64(0.0004481127249057967))

In [37]:
# STEP 5 - Decision rule (α = 0.05)

alpha = 0.05

if p_value < alpha:
    print("Reject the null hypothesis")
else:
    print("Fail to reject the null hypothesis")


Reject the null hypothesis


STEP 6 - INTERPRETATION

A one-sample proportion z-test was performed to determine whether the proportion of hourly workers in the City of Chicago is different from 25%.
The test yielded a z-statistic of −3.51 and a p-value of 0.00045.